### PR TITLE
Refactors telemetry metric registration

### PR DESCRIPTION
**Pull request by Harkishen-Singh**
Signed-off-by: Harkishen-Singh <harkishensingh@hotmail.com>

This commit refactors the metric registration process in telemetry. Now,
the telemetry engine no longer requires to be passed down to each module
that wants to send telemetry, rather just expose a func which accepts a
telemetry engine.

The call of this exposed function happens in runner.go where all modules
that want to send telemetry are called and managed.

Fixes: #1156 

## Merge requirements

Please take into account the following non-code changes that you may need to make with your PR:

- [ ] CHANGELOG entry for user-facing changes
- [ ] Updated the relevant documentation

### CODE REVIEW

File: pkg/runner/runner.go [hunk 1]
```diff
@@ -126,12 +135,27 @@ func Run(cfg *Config) error {
...
+	}
+	if engine == nil {
+		telemetryEngine = telemetry.NewNoopEngine()
+	} else {
+		telemetryEngine = engine
+	}
+	telemetryEngine.Start()
+	defer telemetryEngine.Stop()
+
+	registerMetricsForTelemetry(telemetryEngine)
```
Comment (antekresic) [inline pkg/runner/runner.go:157]: With removal of line 144:
```suggestion
	if engine == nil {
		engine = telemetry.NewNoopEngine()
	} 
	engine.Start()
	defer engine.Stop()

	registerMetricsForTelemetry(engine)
```

File: pkg/api/router.go [hunk 2]
```diff
@@ -116,14 +114,10 @@ func GenerateRouter(apiConf *Config, client *pgclient.Client, elector *util.Elec
...
 	router.Path("/debug/fgprof").Methods(http.MethodGet).HandlerFunc(fgprof.Handler().ServeHTTP)
-
-	if err = registerMetricsForTelemetry(t); err != nil {
-		log.Error("msg", "error registering metrics for telemetry", "err", err.Error())
-	}
 	return router, nil
 }
 
-func registerMetricsForTelemetry(t telemetry.Engine) error {
+func RegisterMetricsForTelemetry(t telemetry.Engine) error {
```
Comment (antekresic) [inline pkg/api/router.go:120]: Naming nit: `RegisterTelemetryMetrics` is shorter and more direct.

**Review by antekresic — APPROVED** [review]
Couple of minor things, great refactortelemetryEngine :+1:

File: pkg/telemetry/telemetry.go [hunk 3]
```diff
@@ -44,13 +44,13 @@ type engineImpl struct {
...
-func NewEngine(conn pgxconn.PgxConn, uuid [16]byte, promqlQueryable promql.Queryable) (*engineImpl, error) {
+func NewEngine(conn pgxconn.PgxConn, uuid [16]byte, promqlQueryable promql.Queryable) (Engine, error) {
 	isTelemetryOff, err := isTelemetryOff(conn)
 	if err != nil {
 		log.Debug("msg", "unable to get TimescaleDB telemetry configuration. Maybe TimescaleDB is not installed", "err", err.Error())
 	}
 	if isTelemetryOff {
-		return nil, nil
+		return NewNoopEngine(), nil
 	} else {
```
Comment (antekresic) [inline pkg/telemetry/telemetry.go:54]: No need for this else since we are returning early.

**Review by antekresic — APPROVED** [review]

File: pkg/tests/end_to_end_tests/telemetry_test.go [hunk 4]
```diff
@@ -32,6 +33,17 @@ func setTobsEnv(prop string) error {
 	return os.Setenv(fmt.Sprintf("TOBS_TELEMETRY_%s", prop), prop)
 }
 
+type e2eTelemetry interface {
```
Comment (niksajakovljevic) [inline pkg/tests/end_to_end_tests/telemetry_test.go:36]: Please add some comments so it's more obvious why is this needed: `we want to manually trigger Sync for tests..`

Comment (niksajakovljevic) [reply pkg/tests/end_to_end_tests/telemetry_test.go:36]: Also consider better naming: eg `testTelemetryEngine`

File: pkg/runner/runner.go [hunk 5]
```diff
@@ -255,3 +261,23 @@ func Run(cfg *Config) error {
...
 }
+
+func registerTelemetryMetrics(client *pgclient.Client) {
+	t, err := telemetry.NewEngine(client.Connection, PromscaleID, client.Queryable())
+	if err != nil {
+		log.Debug("msg", "err creating telemetry engine", "err", err.Error())
+		return
+	}
+	t.Start()
+	defer t.Stop()
```
Comment (niksajakovljevic) [inline pkg/runner/runner.go:272]: This doesn't seem right. Why are we stopping the engine upon telemetry registration?

File: pkg/runner/runner.go [hunk 6]
```diff
@@ -255,3 +261,23 @@ func Run(cfg *Config) error {
 	}
 	return err
 }
+
+func registerTelemetryMetrics(client *pgclient.Client) {
```
Comment (niksajakovljevic) [inline pkg/runner/runner.go:265]: Consider better naming: eg `startTelemetryEngine` since this code is starting the engine and registering few metrics...

File: pkg/runner/runner.go (hunk 5 above)
Comment (niksajakovljevic) [reply pkg/runner/runner.go:272]: I'd recommend returning the engine instance from this function and invoking `Stop` in our shutdown/cleanup stage...

**Review by niksajakovljevic — CHANGES_REQUESTED** [review]
Mainly the telemetry engine lifecycle seems wrong.

File: pkg/runner/runner.go (hunk 5 above)
Comment (Harkishen-Singh) [reply pkg/runner/runner.go:272]: Oh yes, you are right

File: pkg/tests/end_to_end_tests/telemetry_test.go (hunk 4 above)
Comment (Harkishen-Singh) [reply pkg/tests/end_to_end_tests/telemetry_test.go:36]: > Also consider better naming: eg testTelemetryEngine

That name sounds like name for a test function

File: pkg/runner/runner.go (hunk 6 above)
Comment (Harkishen-Singh) [reply pkg/runner/runner.go:265]: Made this `initTelemetryEngine` since starting here only and stopping in caller doesn't look nice. Hence, it initializes and registers metrics and start and deferred stop is done in the caller.

**Review by niksajakovljevic — APPROVED** [review]
